javascript_nodejs/50.contoso-cafe-bot/bot.js

```diff
@@ -0,0 +1,147 @@
+     * 
+     */
+    constructor (conversationState, userState, botConfig) {
+        if(!conversationState) throw ('Missing parameter. conversationState is required');
+        if(!userState) throw ('Missing parameter. userState is required');
+        if(!botConfig) throw ('Missing parameter. botConfig is required');
+
+        // Create state property accessors.
+        this.onTurnPropertyAccessor = conversationState.createProperty(ON_TURN_PROPERTY);
+        this.dialogPropertyAccessor = conversationState.createProperty(DIALOG_STATE_PROPERTY);
+        
+        // add recogizers
+        const luisConfig = botConfig.findServiceByNameOrId(LUIS_CONFIGURATION);
+        if(!luisConfig || !luisConfig.appId) throw (`Cafe Dispatch LUIS model not found in .bot file. Please ensure you have all required LUIS models created and available in the .bot file. See readme.md for additional information\n`);
+        this.luisRecognizer = new LuisRecognizer({
+            applicationId: luisConfig.appId,
+            azureRegion: luisConfig.region,
+            // CAUTION: Its better to assign and use a subscription key instead of authoring key here.
+            endpointKey: luisConfig.authoringKey
+        });
+
+        // add main dialog
+        this.dialogs = new DialogSet(this.dialogPropertyAccessor);
+        this.dialogs.add(new MainDialog(botConfig, this.onTurnPropertyAccessor, conversationState, userState));
+    }
+    /**
+     * On turn dispatcher. Responsible for processing turn input, gather relevant properties,
+     * and continues or begins main dialog.
+     * 
+     * @param {Object} context conversation context object
+     * 
+     */
+    async onTurn (context) {
+        // See https://aka.ms/about-bot-activity-message to learn more about message and other activity types.
+        switch (context.activity.type) {
+            case ActivityTypes.Message: {
+                // Process on turn input (card or NLP) and gather new properties
+                // OnTurnProperty object has processed information from the input message activity.
+                let onTurnProperties = await this.getNewOnTurnProperties(context);
+                if(onTurnProperties === undefined) break;
+                
+                // Set the state with gathered properties (intent/ entities) through the onTurnPropertyAccessor
+                this.onTurnPropertyAccessor.set(context, onTurnProperties);
+                
+                // Do we have any oustanding dialogs? if so, continue them and get results
+                // No active dialog? start a new main dialog
+                await this.continueOrBeginMainDialog(context);
+                break;
+            }
+            case ActivityTypes.ConversationUpdate: {
+                
+                // Send a welcome card to any user that joins the conversation.
+                if(context.activity.membersAdded[0].name !== 'Bot') await this.welcomeUser(context);
```

Comment:
>context.activity.membersAdded[0].name !== 'Bot' [](start = 19, length = 47)

This won't work for every channel

Reply:
@stevenic - do we have a JS sample that shows the right way to do this?